### PR TITLE
Cherrypick "Add some barebones support for DH in EVP" 

### PR DESCRIPTION
**Pull request by samuel40791765**
### Description of changes: 
We have multiple integration projects (Node and Ruby) that happen to depend on this as well. Cherrypicking the support for `EVP_PKEY_DH` over to AWS-LC as well.
There may be more that we need to add to support this better, but this is nice work that helps build something we can start upon. 

### Call-outs:
N/A

### Testing:
Tests in commit

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

**Review by justsmth — APPROVED** [review]

File: include/openssl/evp.h [hunk 1]
```diff
@@ -928,6 +934,22 @@ OPENSSL_EXPORT EVP_PKEY *EVP_PKEY_kem_new_raw_key(int nid,
...
+// EVP_PKEY_CTX_set_dh_pad configures configures whether |ctx|, which must be an
+// |EVP_PKEY_derive| operation, configures the handling of leading zeros in the
+// Diffie-Hellman shared secret. If |pad| is zero, leading zeros are removed
+// from the secret. If |pad| is non-zero, the fixed-width shared secret is used
+// unmodified, as in PKCS #3. If this function is not called, the default is to
+// remove leading zeros.
+//
+// WARNING: The behavior when |pad| is zero leaks information about the shared
+// secret. This may result in side channel attacks such as
+// https://raccoon-attack.com/, particularly when the same private key is used
```
Comment (WillChilds-Klein) [inline include/openssl/evp.h:948]: should we disallow `pad` value of 0?

File: crypto/evp_extra/p_methods.c [hunk 2]
```diff
@@ -14,8 +14,10 @@ static const EVP_PKEY_METHOD *const non_fips_pkey_evp_methods[] = {
   &dilithium3_pkey_meth,
 #endif
   &kem_pkey_meth,
+  &dh_pkey_meth,
```
Comment (WillChilds-Klein) [inline crypto/evp_extra/p_methods.c:17]: why not FIPS?

Comment (WillChilds-Klein) [reply crypto/evp_extra/p_methods.c:17]: ah, got it.

File: crypto/fipsmodule/evp/evp.c [hunk 3]
```diff
@@ -425,19 +422,9 @@ EC_KEY *EVP_PKEY_get1_EC_KEY(const EVP_PKEY *pkey) {
   return ec_key;
 }
 
-DH *EVP_PKEY_get0_DH(const EVP_PKEY *pkey) {
-  SET_DIT_AUTO_DISABLE;
-  return NULL;
-}
-
-DH *EVP_PKEY_get1_DH(const EVP_PKEY *pkey) {
-  SET_DIT_AUTO_DISABLE;
```
Comment (WillChilds-Klein) [inline crypto/fipsmodule/evp/evp.c:434]: hm, this isn't touching secret material. we didn't need this call here previously, did we?

**Review by WillChilds-Klein — DISMISSED** [review]

File: crypto/fipsmodule/evp/evp.c (hunk 3 above)
Comment (samuel40791765) [reply crypto/fipsmodule/evp/evp.c:434]: Really good point, let me ask internally about the justification behind this and if I should add this back.

**Review by WillChilds-Klein — DISMISSED** [review]

**Review by justsmth — DISMISSED** [review]

**Review by justsmth — DISMISSED** [review]

**Review by WillChilds-Klein — DISMISSED** [review]

**Review by smittals2 — DISMISSED** [review]

**Review by WillChilds-Klein — APPROVED** [review]

**Review by WillChilds-Klein — APPROVED** [review]

**Review by smittals2 — APPROVED** [review]

**Pull request merged**